### ISSUE TITLE
Bulk import performance

### ISSUE THREAD

**Issue by IanMayo** [debrief/pepys-import#45]
An attached zip-file contains a high volume of data-files in REPLAY format.

We should use the sample data-set to investigate Pepys-Import performance bottlenecks, and see where performance can be improved.
[BulkAssetRuns.zip](https://github.com/debrief/pepys-import/files/3846385/BulkAssetRuns.zip)

**Comment by IanMayo** [comment]
We have a one or two really large files in our sample data. I can get others.  It would be good if you could get some metrics on how long they take to import - maybe broken down by what’s happening at each stage (parse vs insert vs extraction highlighter).

Then as you address any bottlenecks we will see the performance improvement.

It's most important that we perform well in PG, but I am interested in SQLite performance too.

**Comment by robintw** [comment]
Just FYI @BarisSari - you can turn off the extraction highlighter for a particular importer by calling `self.disable_recording()` in the `__init__` method of that importer. That will mean the extraction highlighter won't create its internal arrays, and it means that the `record` calls with become a no-op. Will be interesting to see how that speeds things up for large files.

**Comment by BarisSari** [comment]
Thanks, @robintw . I'm trying to profile the import for using cProfile. It seems like that the highlighter takes at least 5-6% of the time. I'll import again with disabling recording as well. Because I'm sure that the highlighter calls many built-in functions as well. So, it is probably consuming more than 5-6 percent.

**Comment by BarisSari** [comment]
First, I created a basic script,` local_file.py` and I profiled the following options using this script:
```python
from pepys_import.core.store.data_store import DataStore
from importers.replay_importer import ReplayImporter
from pepys_import.file.file_processor import FileProcessor

DATA_PATH = "/Users/baris/Documents/PycharmProjects/pepys-import/tests/sample_data/results2"  # bulk data folder

data_store = DataStore("", "", "", 0, "test.db", db_type="sqlite")
data_store.initialise()

processor = FileProcessor()
processor.register_importer(ReplayImporter())
processor.process(DATA_PATH, data_store, False)
```

- Run bulk-import without modifying the code
![pepys-import10](https://user-images.githubusercontent.com/41292272/78771343-b2cd2480-7998-11ea-9c40-d17446492ae2.png)

- Run bulk-import disabling the extraction highlighter
![pepys-import12_without_highlighter](https://user-images.githubusercontent.com/41292272/78771230-86b1a380-7998-11ea-8126-0d483a7ef638.png)

- Run bulk-import disabling the extraction highlighter and removing assert statements from the `get_datafile`, `get_platform` and `get_sensor` methods.

Actually, this one changed almost nothing. I thought `built-in isinstance()` is called too many times because of our methods but it's not. It didn't change the performance at all. That's why I'm not putting its screenshot.

- Run bulk-import disabling the extraction highlighter and removing tqdm(progress bar)
![pepys-import14_w_out_tqdm](https://user-images.githubusercontent.com/41292272/78773011-456ec300-799b-11ea-9b46-9cddd2a58eed.png)

`Tqdm` also affected only a little. Execution time decreased 5 s which is ~0.0003 of the total execution time. So, I believe that it's negligible.

I didn't cut the graphs because you can realise something that I didn't realise. If you do, please point out. I think you can start looking at the graphs from the `local_file.py`->`process`->`process_file`. Then, you will see that it has branches that represent parsing (`load_this_file`) and inserting (`validate` & `commit`).

My comment in overall is as follows:
It seems that we don't have any bottlenecks or erroneous methods/functions which consume lots of time. The only problematic thing we have is the extraction highlighter. It consumes approximately 8-9% of the total execution time(If you subtract `process_files`' total milliseconds of the first two graphs, you can see the difference). I'm not familiar with it because I never touched this part of the repository. Maybe @robintw can have some suggestions to improve it. However, I believe that we can improve it (at least we can change the `color_picker.py` and check the performance again). Maybe using `Numpy` or `math` could improve the performance a bit.

**Note:** You probably expect to see export_report and related methods in the first graph. However, I couldn't extract it and I don't know why. Yet, I have a table of functions/methods, their call counts, their total time and their own time. I can share it with you if you would like to see it.
**Note-2:** I tried only on SQLite for now. I'll tests it on PostgreSQL as well. If it takes too much time, I think I'll test it with only 2-3 REPL files.

I hope that this analysis is a good start to discuss performance improvements. What are your thoughts, @IanMayo, and @robintw ?

**Comment by IanMayo** [comment]
It has taken me a few minutes to understand what I'm looking at.

I was _expecting_ this review to identify the opportunity for us to do a bulk SQL INSERT (rather than one at a time).  This _normally_ brings performance improvements.

**Comment by BarisSari** [comment]
> It has taken me a few minutes to understand what I'm looking at.
> 
> I was _expecting_ this review to identify the opportunity for us to do a bulk SQL INSERT (rather than one at a time). This _normally_ brings performance improvements.

Sorry for the misunderstanding. I thought you meant to run our code over the REP files in the zip and check the performance of it. I'll test bulk SQL Insert and share it.

**Comment by BarisSari** [comment]
Okay, I have searched for the bulk-import options. SQL Alchemy offers a few different options such as ORM's `bulk_save_objects`-`bulk_insert_mappings`, Core's `engine.connect.execute`. The fastest one is using the core but it is a problem to use it for us because we are opening one session for the import process and we can't open a connection inside of a session (It locks the database and throws an error). So, if we would like to use it, we should change our session approach. `bulk_insert_mappings` accepts a dictionary, and `bulk_save_objects` accepts DB objects. Since we are creating intermediary DB objects, `bulk_save_objects` looks like the best option. When I changed our commit method:
```python
        for parser in self.measurements:
            total_objects = 0
            for platform, objects in self.measurements[parser].items():
                total_objects += len(objects)
                print(f"Submitting measurements extracted by {parser}.")
                # Bulk save table objects; state, etc.
                data_store.session.bulk_save_objects(objects)
```
  The performance of the import increased by ~17% (It is nearly the same with the execution time of `submit()`). However, in this format, we don't create and Log entities for each created object and we don't expire the location field of the objects(which were necessary to have a proper element instead of string). 

I also did one more change to use psycopg2's fast execution helpers. It is in `data_store.py`:
```python
        if db_type == "postgres":
            self.engine = create_engine(connection_string, echo=False, executemany_mode="batch")
            BasePostGIS.metadata.bind = self.engine
        elif db_type == "sqlite":
            self.engine = create_engine(connection_string, echo=False)
            listen(self.engine, "connect", load_spatialite)
            BaseSpatiaLite.metadata.bind = self.engine
```

`executemany_mode` parameter offers the following:

> 'batch' - Uses psycopg2.extras.execute_batch so that multiple copies of a SQL query, each one corresponding to a parameter set passed to executemany(), are joined into a single SQL string separated by a semicolon. This is the same behavior as was provided by the use_batch_mode=True flag.

So, it will be useful if we decide to use `engine.connect.execute()`.

I'd like to share a few links so that you can make further reading if you are interested in :
https://docs.sqlalchemy.org/en/13/faq/performance.html#i-m-inserting-400-000-rows-with-the-orm-and-it-s-really-slow (Be aware that it's in python2 but has valuable information)
https://docs.sqlalchemy.org/en/13/core/tutorial.html#executing-multiple-statements

**Comment by IanMayo** [comment]
I'd appreciate @robintw taking a look at @BarisSari 's suggestions - and also having a look at the links, to maintain his "technical lead" overview.

**Comment by robintw** [comment]
A useful tidbit that was handy when investigating this and putting together #330. There is a plugin for pytest called `pytest-profiling`. Install it with `pip install pytest-profiling` and then you can run a test with profiling by running:

```
tests/highlighter_tests/test_combine.py::CombineTokenTests::test_combine_lines_on_multiple_lines--profile-svg
```

This will run that specific test, print out a summary of the profile, store a .prof file for more detailed investigation, and produce a SVG coloured flowchart showing the functions that take the most time. Very handy!

**Comment by IanMayo** [comment]
Thanks for that @robintw .  Do you know of a technique/technology that would allow us to force a build to fail if we drop beneath some performance threshold?   Maybe it's something that would be configured via GitHub, or maybe it would just be a dedicated Python Unit test.

This would give us _belt and braces_ against introducing some feature that has an unexpected performance cost.

**Comment by robintw** [comment]
There is a [pytest-benchmark](https://pytest-benchmark.readthedocs.io/en/latest/index.html) plugin which allows various things to be done regarding benchmarks in tests. There are two ways we could use this:

1. Use some code like [this](https://github.com/ionelmc/pytest-benchmark/issues/40) to make a test fail if it takes longer than a certain time. This would be a fixed time (eg. 10 seconds) and would not vary depending on the computer it was run on, and wouldn't compare to previous runs.

2. Use the plugin's support for [comparing previous runs](https://pytest-benchmark.readthedocs.io/en/latest/comparing.html) to automatically cause the tests to fail if the time taken increases by more than a certain amount (eg. 10% or 10 seconds). This would be more complicated to implement though, as we'd need to keep a separate set of previous benchmark runs for each machine that we've run it on. We should be able to do this for Travis using the [Travis caching support](https://docs.travis-ci.com/user/caching/), and we could just gitignore the directory containing the previous runs on our local machines.

Which do you think sounds better?

**Comment by IanMayo** [comment]
Hello @robintw - I think we'll be ok with the first one, to start with.

We can come up with some "reasonable" time for an import to work, and compare against that. That should show us if some modification has a drastic impact on the import performance.

It would be wrong for the test to run through all the sample data - since that will add an extra 10 mins or so to the build time.   But, we need enough data to "exercise" the import process.  Aah, our test format should also include extraction highlighting.

I've attached a REP file that contains state/contact/comment data - so it drives 3 importers.
[bulk_data.rep.zip](https://github.com/debrief/pepys-import/files/4476120/bulk_data.rep.zip)

When you do come to work on the performance "regression" test, I'd appreciate you unpacking the zip, and including the file in the `track_files\rep_data\` folder, please.

**Comment by robintw** [comment]
Thanks Ian. Did you want me to include that file just so that we have a file that drives those three importers? If so, I think `rep_test1.rep` does that already - although it's much shorter so won't be so good for testing import performance.

**Comment by IanMayo** [comment]
Yes please @robintw  - I generated it this afternoon, aiming to have something long enough to produce meaningful timings.

**Comment by robintw** [comment]
Fixed in #335, #330, and #328

**Comment by IanMayo** [comment]
I'm going to re-open this, just to let us double-check the performance against an even larger dataset - in case there are bottlenecks that only appear when the import is _really_ pushed.

This Debrief issue contains zip files large volumes of track data: 
https://github.com/debrief/debrief/issues/4849#issuecomment-617827482

We should unpack them into one folder, then run pepys-import on that folder. It would probably be useful to verify it against SQLite and Postgres.  The `anywheredata` hosting has a 100Mb limit, but the SQLite version of this data is at least 300Mb.  So, we can probably only test it against a local PostGres database.  (plus, in real life, the data won't be pushed to the database over an Internet connection).

**Comment by robintw** [comment]
Just a brief note (I'm not officially working today) to say that I've found a way to double the speed (halve the time taken for an import) after a bit more profiling and inspection of the code. I'm 95% sure that the optimisation I'm doing is valid, but need to check a bit more tomorrow and then submit a PR for you both to review. Good news though!

**Comment by IanMayo** [comment]
Issue closed (for now) following #342 and #330
